typing/typedtree.ml

```diff
@@ -480,10 +480,18 @@ and value_description =
     val_attributes: attribute list;
     }
 
+and type_parameter =
+  {
+    typa_type: type_expr;
+    typa_name: string option loc;
```

Comment:
Before we had a `core_type`, and now we only have a `type_expr` and a `string option loc`. Your description suggests that you are just turning a tuple into a record, but here we change the representation in a deeper way.

(My understanding is that `core_type` corresponds to the typed representation of a "syntactic type", that syntactically occurs in the program, and that `type_expr` is the internal representation of the type in the type-checker.)

For example, this changes the implementation of Tast_iterator. Tast_mapper. Before, the `typ` method (that operates on `core_type ) would get called on type variables in parameter position. After the PR, this part of the visitor is not called anymore on parameters (in fact the parameters are not traversed at all). I don't know if the change is a good thing or not. (Are there a generic traversal that one could define that becomes less correct, or more correct, with this change?)

Reply:
As a data point @nchataing wrote a traversal as part of his work on migrating LexiFi's codebase to use `floatarray` that I believe makes use of this.

@nchataing can you confirm?

Reply:
This field used to be a `string option loc`, and was changed to a core_type in (I think?) 4.02. My understanding is that the  motivation for the change in Typedtree was to make a `type_expr` for the parameter available, which at the time was done by making it a `core_type`, which already had one.

I'm struggling to imagine a use-case of Tast_iterator that breaks with this change. The type parameters are variable binders, but Tast_iterator.typ does not specify whether a variable occurrence is a binder or a use or what scope it might be in, so I don't see how it's possible to make use of the current behaviour. (I might be missing something here, so I am curious about @nchataing's work, though!)

Reply:
@nchataing tells me that I am mistaken and that this is unrelated to what is used in his traversal. Sorry for the noise.